core/embed/extmod/modtrezorcrypto/modtrezorcrypto-secp256k1_zkp.h

```diff
@@ -325,13 +331,265 @@ STATIC mp_obj_t mod_trezorcrypto_secp256k1_context_multiply(
                       secp256k1_ecdh_hash_passthrough, NULL)) {
     mp_raise_ValueError("Multiply failed");
   }
+  // Verify and re-serialize into (optionally compressed) public key.
+  size_t outlen = sizeof(out);
+  if (!secp256k1_ec_pubkey_parse(ctx, &ec_pk, out, outlen)) {
+    mp_raise_ValueError("Invalid shared public key");
+  }
+  secp256k1_ec_pubkey_serialize(
+      ctx, out, &outlen, &ec_pk,
+      compressed ? SECP256K1_EC_COMPRESSED : SECP256K1_EC_UNCOMPRESSED);
+
+  return mp_obj_new_bytes(out, outlen);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(
+    mod_trezorcrypto_secp256k1_context_multiply_obj, 3, 4,
+    mod_trezorcrypto_secp256k1_context_multiply);
+
+/// def blind_generator(asset: bytes, blinding_factor: bytes) -> bytes:
+///     '''
+///     Generate blinded generator for the specified confidential asset.
+///     '''
+STATIC mp_obj_t mod_trezorcrypto_secp256k1_context_blind_generator(
+    mp_obj_t self, mp_obj_t asset_obj, mp_obj_t blind_obj) {
+  const secp256k1_context *ctx = mod_trezorcrypto_get_secp256k1_context(self);
+
+  mp_buffer_info_t asset;
+  mp_get_buffer_raise(asset_obj, &asset, MP_BUFFER_READ);
+  if (asset.len != 32) {
+    mp_raise_ValueError("Invalid length of asset");
+  }
+  mp_buffer_info_t blind;
+  mp_get_buffer_raise(blind_obj, &blind, MP_BUFFER_READ);
+  if (blind.len != 32) {
+    mp_raise_ValueError("Invalid length of blinding factor");
+  }
+  secp256k1_generator gen;
+  if (!secp256k1_generator_generate_blinded(ctx, &gen, asset.buf, blind.buf)) {
+    mp_raise_ValueError("Generator blinding failed");
+  }
+  byte out[33] = {0};
+  secp256k1_generator_serialize(ctx, out, &gen);
   return mp_obj_new_bytes(out, sizeof(out));
 }
+
 STATIC MP_DEFINE_CONST_FUN_OBJ_3(
-    mod_trezorcrypto_secp256k1_context_multiply_obj,
-    mod_trezorcrypto_secp256k1_context_multiply);
+    mod_trezorcrypto_secp256k1_context_blind_generator_obj,
+    mod_trezorcrypto_secp256k1_context_blind_generator);
+
+// TODO: this code is copied from modtrezorcrypto-monero.h - find a better way
+// to handle the conversion.
+static uint64_t _mp_obj_get_uint64(mp_const_obj_t arg) {
+  if (MP_OBJ_IS_SMALL_INT(arg)) {
+    return MP_OBJ_SMALL_INT_VALUE(arg);
+  } else if (MP_OBJ_IS_TYPE(arg, &mp_type_int)) {
+    byte buff[8];
+    uint64_t res = 0;
+    mp_obj_t *o = MP_OBJ_TO_PTR(arg);
+
+#if MICROPY_LONGINT_IMPL != MICROPY_LONGINT_IMPL_MPZ
+#error "MPZ supported only"
+#endif
+    mp_obj_int_to_bytes_impl(o, true, 8, buff);
+    for (int i = 0; i < 8; i++) {
+      res <<= i > 0 ? 8 : 0;
+      res |= (uint64_t)(buff[i] & 0xff);
```

Comment:
I wonder whether this is endian-dependent (if yes, should maybe be documented or #error out if wrong)

Reply:
IIUC `mp_obj_int_to_bytes_impl` should order the bytes according to its 2nd argument:
https://github.com/micropython/micropython/blob/b9ec6037edf5e6ff6f8f400d70f7351d1b0af67d/py/objint.h#L58
I have updated the invocation at https://github.com/romanz/trezor-firmware/pull/1/commits/729918821371de7462160ddc5e18f55944e26871 to make it more explicit.

Reply:
Refactored at b185e8475d22f05531eca0ec410d9cbd26a333e5.